src/vm/tests/large_contract.rs

```diff
@@ -0,0 +1,103 @@
+
+    let contract_identifier = QualifiedContractIdentifier::local("foo").unwrap();
+
+    {
+        let mut conn = clarity_instance.begin_block(&TrieFileStorage::block_sentinel(),
+                                                    &BlockHeaderHash::from_bytes(&[0 as u8; 32]).unwrap(),
+                                                    &NULL_HEADER_DB);
+
+        let define_data_var = "(define-data-var XZ (buff 1048576) \"a\")";
+
+        let mut contract = define_data_var.to_string();
+        for i in 0..20 {
+            let cur_size = format!("{}", 2u32.pow(i));
+            contract.push_str("\n");
+            contract.push_str(
+                &format!("(var-set XZ (concat (unwrap-panic (as-max-len? (var-get XZ) u{}))
+                                             (unwrap-panic (as-max-len? (var-get XZ) u{}))))",
+                        cur_size, cur_size));
+        }
+        for i in 0..EXPLODE_N {
+            let exploder = format!("(define-data-var var-{} (buff 1048576) (var-get XZ))", i);
+            contract.push_str("\n");
+            contract.push_str(&exploder);
+        }
+
+        let (ct_ast, ct_analysis) = conn.analyze_smart_contract(&contract_identifier, &contract).unwrap();
+        assert!(format!("{:?}",
+                        conn.initialize_smart_contract(
+                            &contract_identifier, &ct_ast, &contract, |_,_| false).unwrap_err())
+                .contains("MemoryBalanceExceeded"));
+    }
+}
+
+/*
+ */
+#[test]
+pub fn let_memory_test() {
+    let marf = MarfedKV::temporary();
+    let mut clarity_instance = ClarityInstance::new(marf);
+    let EXPLODE_N = 100;
+
+    let contract_identifier = QualifiedContractIdentifier::local("foo").unwrap();
+
+    {
+        let mut conn = clarity_instance.begin_block(&TrieFileStorage::block_sentinel(),
+                                                    &BlockHeaderHash::from_bytes(&[0 as u8; 32]).unwrap(),
+                                                    &NULL_HEADER_DB);
+
+        let define_data_var = "(define-constant buff-0 \"a\")";
+
+        let mut contract = define_data_var.to_string();
+        for i in 0..20 {
+            contract.push_str("\n");
+            contract.push_str(
+                &format!("(define-constant buff-{} (concat buff-{} buff-{}))",
+                         i+1, i, i));
+        }
+
+        contract.push_str("\n");
+        contract.push_str("(let (");
+
+        for i in 0..EXPLODE_N {
+            let exploder = format!("(var-{} buff-20) ", i);
+            contract.push_str(&exploder);
+        }
+
+        contract.push_str(") 1)");
+
+        let (ct_ast, ct_analysis) = conn.analyze_smart_contract(&contract_identifier, &contract).unwrap();
+        assert!(format!("{:?}",
+                        conn.initialize_smart_contract(
+                            &contract_identifier, &ct_ast, &contract, |_,_| false).unwrap_err())
+                .contains("MemoryBalanceExceeded"));
+    }
+}
```

Comment:
Do we also want tests for:
* Very long function arguments? (or functions with very long argument lists)?
* Very deep functions?
* Very big contracts?
* Tuples, maps, data vars, and NFTs with very large type signatures?

Reply:
If you don't want to address them here, then please add an issue for it.  I think it's going to be important that we extensively test this PR and other resource-limiting code.

Reply:
> Very long function arguments? (or functions with very long argument lists)?

I'm still not sure how and when function arguments should be bounded -- I think this limit has more to do with validating transactions for the mempool than any Clarity interpreter constraints. At the moment, the bound applied is just the memory limit -- i.e., if the argument list exceeds maximum memory limits. I can add a test for that here pretty simply.

> Very deep functions?

That's tested a number of different ways elsewhere (AST depth limits for defining very deep functions, call depth limits, etc.). Or do you mean a test specifically for memory limits in nested calls?

> Very big contracts?

Clarity doesn't enforce a limit on contract sizes, that's enforced via transaction size limits.

> Tuples, maps, data vars, and NFTs with very large type signatures?

Too large type signatures are tested in a variety of ways in the type module's tests.

Reply:
Tests for function long arguments exceeded memory limit and nested function calls exceeding the memory limit added in df99ce264

Reply:
> Or do you mean a test specifically for memory limits in nested calls?
Yes, this.  The amount of available memory in a deeper context ought to be smaller than the amount of available memory in a shallower context, since each additional context consumes memory.

> Clarity doesn't enforce a limit on contract sizes, that's enforced via transaction size limits.

Doesn't an executing transaction consume memory if it loads a contract in the process?  i.e. to do a contract-call?

Reply:
> Doesn't an executing transaction consume memory if it loads a contract in the process? i.e. to do a contract-call?

Yes -- I added a contract-call test in `8b4e410f6`

Reply:
Thanks!  Sorry for the poor choice of words on my part.  Meant "large" to mean "takes a lot of memory to process" not "takes a lot of memory to represent."